Symlink in the exec-root so that relative paths will work, unchanged.

freeformstu opened the pull request:
#1762

illicitonion commented on cargo/cargo_build_script_runner/bin.rs:
```diff
@@ -53,6 +53,46 @@ fn run_buildrs() -> Result<(), String> {
+            )
+        })?;
+
+        // Only symlink directories.
```
Why not symlink files too?

freeformstu replied:
There's actually no reason not to symlink in files as well. I did not need it for my use case so I ignored it. I'll get file symlinks added in.

illicitonion reviewed (commented):
This is a pretty cute solution, thanks for putting it together!

Could you add add a test-case to https://github.com/bazelbuild/rules_rust/tree/main/test/cargo_build_script which shows that this works, so that we don't break it in the future?

illicitonion commented on cargo/cargo_build_script_runner/bin.rs:
```diff
@@ -53,6 +53,30 @@ fn run_buildrs() -> Result<(), String> {
+
+        let file_name = path
+            .file_name()
+            .ok_or_else(|| "Failed while getting file name".to_string())?
+            .to_str()
+            .ok_or_else(|| "Failed while converting file name to string".to_string())?;
```
```suggestion
            .ok_or_else(|| "Failed while getting file name".to_string())?;
```

You can use an `OsString` to `Path::join`

illicitonion commented on test/cargo_build_script/BUILD.bazel:
```diff
@@ -66,3 +66,40 @@ rust_test(
+###############################################################################
+
+write_file(
+    name = "root_file",
+    out = "root_file.txt",
```
```suggestion
    name = "cargo_manifest_dir_file",
    out = "cargo_manifest_dir_file.txt",
```

My first read of this in the build script assumed "root" meant "root of repo"

illicitonion commented on test/cargo_build_script/test_exec_root_access.build.rs:
```diff
@@ -0,0 +1,62 @@
+        "Build script must be in the current directory"
+    );
+    assert!(
+        root_files.take("root_file.txt").is_some(),
```
Let's assert on the contents too

illicitonion reviewed (approved):
LGTM, just some minor comments - thanks!

freeformstu commented:
@illicitonion Apologies for the delay on this PR. I just addressed your comments - I'm still working through some windows test failures in this PR. I'll try to get those addressed by the end of this week.

Do you mind if I squash and rebase?

illicitonion commented:
Thanks! Feel free to manage the branch however you want (rewrite history or add commits on top), and I'll squash + rebase into main whenever it's done :)

UebelAndre commented on cargo/cargo_build_script_runner/bin.rs:
```diff
@@ -53,6 +53,23 @@ fn run_buildrs() -> Result<(), String> {
         .unwrap_or_else(|_| panic!("Failed to make output directory: {:?}", out_dir_abs));
 
+    // Symlink the execroot to the manifest_dir so that we can use relative paths in the arguments.
+    let exec_root_paths = std::fs::read_dir(&exec_root)
```
Ok, I think the issue here is that when running outside of a sandbox there will actually be more files in this directory than expected. Instead of reading from the execroot, is there a MANIFEST file or something that tracks all inputs to an action? I forget if that's created for actions or just tests. The idea would be you can link only the things the build script has as an explicitly registered as an input.

UebelAndre replied:
Looks like MANIFEST is only found in runfiles

freeformstu replied:
Thanks @UebelAndre! I can reproduce the issue locally with this diff now:
```diff
diff --git a/cargo/private/cargo_build_script.bzl b/cargo/private/cargo_build_script.bzl
index e567c1c5..3d1f4858 100644
--- a/cargo/private/cargo_build_script.bzl
+++ b/cargo/private/cargo_build_script.bzl
@@ -242,6 +242,9 @@ def _cargo_build_script_impl(ctx):
         mnemonic = "CargoBuildScriptRun",
         progress_message = "Running Cargo build script {}".format(pkg_name),
         env = env,
+        execution_requirements = {
+            "local": "",
+        }
     )
 
     return [
```

What do you think the behavior should be in this case? Should I just disable all of this logic if you're building outside of a bazel sandbox since that's the whole point of this change?

freeformstu replied:
My most recent change updates the test to only assert that the execroot is empty if the test is running in a sandbox. This resolves almost all of the issues but now I have 3 remaining. One is definitely symlink related and the other two look unrelated but may be

Questions:
1. Would it be acceptable to make this a feature which does not exist when running on Windows so that it can entirely avoid the windows symlinking issues?
2. If that's not acceptable, would it be acceptable to make this feature opt-in (perhaps through an environment variable) so that consumers of this change can toggle it on where they see fit? That would give folks the ability to enable it for windows in the environments where they have symlinks enabled.

UebelAndre replied:
I think making this a feature would be good. It can be disabled by default and enabled globally via `--features` in user's `.bazelrc` files. My experience is doing anything on windows is a bad idea 😅 but we should probably aim to keep the behavior consistent across operating systems.

UebelAndre reviewed (changes requested):
I think the failure here is caused by the Windows not actually having support for sandboxing. I don't have a windows box so it's kinda hard to confirm but reading https://bazel.build/versions/6.1.0/docs/sandboxing it seems to explicitly mention Linux and MacOS but not Windows which, in addition to the failure here, makes me think there's just no sandboxing. This is potentially a good find for the feature as we should make sure it works correctly with and without sandboxing.

freeformstu commented:
Now I just have ios and mac build issues which look like they're just generally broken (not due to my changes). @UebelAndre any recommendations here?

UebelAndre commented on .bazelci/presubmit.yml:
```diff
@@ -27,6 +27,7 @@ default_windows_targets: &default_windows_targets
   - "-//bindgen/..."
   - "-//test/proto/..."
   - "-//test/unit/pipelined_compilation/..."
+  - "-//test/cargo_build_script:test_exec_root_access.build.feature_enabled"
```
Can you delete this in favor of a `target_compatible_with` flag on the target should you need it?

UebelAndre commented on cargo/private/cargo_build_script.bzl:
```diff
@@ -3,6 +3,7 @@
 load("@bazel_skylib//lib:paths.bzl", "paths")
 load("@bazel_tools//tools/build_defs/cc:action_names.bzl", "CPP_COMPILE_ACTION_NAME", "C_COMPILE_ACTION_NAME")
 load("@bazel_tools//tools/cpp:toolchain_utils.bzl", "find_cpp_toolchain")
+load("//cargo:features.bzl", "SYMLINK_EXEC_ROOT_FEATURE", "feature_enabled")
```
Why would this need to be a part of the public api vs living in `//cargo/private`?

UebelAndre replied:
That sounds reasonable, but could you then move the `feature_enabled` logic into `//cargo/private`? Surely that doesn't seem necessary, right?

freeformstu replied:
Done.

UebelAndre reviewed (changes requested):
Awesome work! Just a few questions 😅

UebelAndre reviewed (approved):
Amazing work! Thank you so much!

dk333333 commented:
Hi @freeformstu apologies for the naive question, but how/where do I enable this feature? I had cherry picked an earlier commit but will need to enable it for real when upgrading. I checked the [docs](http://bazelbuild.github.io/rules_rust/flatten.html) but could not find it there
